### DISCUSSION TITLE
How to apply distort function to fabricjs canvas ?

### DISCUSSION THREAD

**Discussion by hoangtd185** [fabricjs/fabric.js#8638] [Q&A]
Hello guys,

So I want to apply a distort function like fisheyejs in fabric 
(example here: http://ericleong.me/fisheye.js/)

Or bulge/pinch like glfx.js (Example here: https://evanw.github.io/glfx.js/demo/#bulgePinch)
<img width="1038" alt="image" src="https://user-images.githubusercontent.com/32231913/216243104-1bdffd2c-d21e-4483-a4d0-89a23e89daab.png">

But those library can't be used with fabric. So is there any other library can  work with fabricjs and has that function ?

Or the only way is create a custom filter for it ? 

I'm  stuck at this task  for a  week but can't find solution. So please help me. Thank you so much

**Comment by ShaMan123** [comment]
Reading a bit the repos you referenced I am not sure you are right stating they can't work with fabric.
Seems both can be applied on img/canvas so it might be possible to subclass Image and apply those functions at a correct time and then render canvas...
Also it might be possible to apply the same on the underlying canvas element (`canvas.getElement()`), perhaps in the `after:render` hook.

It would be great if you provided a demo or PR after you managed doing it.

And of course the fabric way to do it seems to be a filter. I am not experienced in that field. But you can use the webGL logic from the repos...

**Comment by ShaMan123** [comment]
Looks like it can be done pretty straight forward with a filter:

https://github.com/evanw/glfx.js/blob/58841c23919bd59787effc0333a4897b43835412/src/filters/warp/bulgepinch.js
https://github.com/evanw/glfx.js/blob/86ae72acfffcd50774507eb7b69887ba2493dc35/src/core/canvas.js#L122
https://github.com/evanw/glfx.js/blob/86ae72acfffcd50774507eb7b69887ba2493dc35/src/core/canvas.js#L85


https://github.com/evanw/glfx.js/blob/86ae72acfffcd50774507eb7b69887ba2493dc35/src/core/canvas.js#L132-L133
As you can see the exported canvas creates a canvas element. You can pass that to fabric `new fabric.Canvas(el)` and see what's next
